File: packages/orchestrator/src/orchestrator.ts
import { EventEmitter } from "events";
import { randomUUID } from "crypto";
import { createDoubaoClient } from "./llm/doubao.js";
import { createConduitRepo } from "./repo/conduit.js";
import { emit as storeEmit, truncateAfter, readEvents } from "./events/store.js";
import { persistMemory, readAllMemories } from "./memory/store.js";
import { extractEntities } from "./memory/extract.js";
import { recall } from "./memory/recall.js";
import { validateRecalledFiles } from "./memory/validate.js";
import { loadSkills } from "./skills/registry.js";
import { aspectScan, filterRecallForScan } from "./agents/aspect-scan.js";
import { buildQuestionsFromAspects } from "./agents/clarify.js";
import { clarifyAgent } from "./agents/clarify.js";
import { planAgent } from "./agents/plan.js";
import { locateAgent } from "./agents/locate.js";
import { codeAgent } from "./agents/code.js";
import { verifyAgent } from "./agents/verify.js";
import type { RunEvent, ClarifiedRequest, ChangeSet, Skill } from "./types.js";

const MAX_VERIFY_RETRIES = 2;

// ────────────────────────────────────────────────────────────
// 单例事件总线，API 层通过它订阅特定 runId 的事件
// ────────────────────────────────────────────────────────────
export const bus = new EventEmitter();
bus.setMaxListeners(100);

function emitAndBroadcast<T>(runId: string, type: RunEvent["type"], payload: T): RunEvent<T> {
  const event = storeEmit(runId, type, payload);
  bus.emit(`run:${runId}`, event);
  return event;
}

// ────────────────────────────────────────────────────────────
// 暂停/恢复 + 中断
// ────────────────────────────────────────────────────────────

type AnswerResolve = (answers: Record<string, string>) => void;

const pendingAnswers = new Map<string, AnswerResolve>();
// 记录每个 run 当前所处阶段，供 replay 使用
const runPhase = new Map<string, string>();

export function provideClarificationAnswers(runId: string, answers: Record<string, string>): boolean {
  const resolve = pendingAnswers.get(runId);
  if (!resolve) return false;
  pendingAnswers.delete(runId);
  resolve(answers);
  return true;
}

function waitForAnswers(runId: string): Promise<Record<string, string>> {
  return new Promise((resolve) => pendingAnswers.set(runId, resolve));
}

// ────────────────────────────────────────────────────────────
// PR 描述生成
// ────────────────────────────────────────────────────────────

async function buildPRDescription(runId: string, req: ClarifiedRequest, diff: string): Promise<string> {
  const llm = createDoubaoClient({ runId });
  const result = await llm.chat([
    {
      role: "system",
      content: "根据需求描述和 git diff，输出 PR 标题（第一行）和描述（Markdown），不超过 200 字。",
    },
    {
      role: "user",
      content: `需求：${req.summary}\n\n字段：${req.fieldName}\n\n规则：${req.businessRule}\n\ndiff（前100行）：\n${diff.split("\n").slice(0, 100).join("\n")}`,
    },
  ], undefined, { agent: "pr:description" });
  return result.text;
}

// ────────────────────────────────────────────────────────────
// 主编排流程
// ────────────────────────────────────────────────────────────

export async function startRun(rawText: string): Promise<string> {
  const runId = randomUUID();
  const llm = createDoubaoClient({ runId });
  const repo = createConduitRepo();

  // 异步跑，不阻塞调用方
  runPipeline(runId, rawText, llm, repo).catch((err) => {
    emitAndBroadcast(runId, "run.error", { message: String(err?.message ?? err) });
    runPhase.delete(runId);
  });

  return runId;
}

async function runPipeline(
  runId: string,
  rawText: string,
  llm: ReturnType<typeof createDoubaoClient>,
  repo: ReturnType<typeof createConduitRepo>
): Promise<void> {
  emitAndBroadcast(runId, "run.started", {
    rawText,
    recallDisabled: process.env.DISABLE_RECALL === "true",
  });
  runPhase.set(runId, "clarify");

  // ── 1. Clarify analyze（只解析语义，不生成问题）──────────
  const allSkills = await loadSkills();
  const aspectUnion = Array.from(new Set(allSkills.flatMap((s) => s.possibleAspects)));
  const analysis = await clarifyAgent.analyze(rawText, llm);

  // ── 1.5. Recall（基于 partial.summary，让 RecallCard 早于 ClarifyBox）──
  let locateAttempt = 1;
  let recallMatchesForScan: Parameters<typeof filterRecallForScan>[0] = [];
  const recallDisabled = process.env.DISABLE_RECALL === "true";
  if (recallDisabled) console.log(`[recall] DISABLE_RECALL=true，跳过召回（baseline 模式）`);
  if (!recallDisabled) try {
    const queryEntities = await extractEntities(llm, {
      runId,
      summary: analysis.req.summary,
      skillUsed: "",
    });
    const memories = readAllMemories();
    const matches = recall(queryEntities, memories, { topK: 3, minScore: 0.3 });

    if (matches.length > 0) {
      const top = matches[0];
      const matchesPayload = matches.map((m) => ({
        runId: m.memory.runId,
        score: m.score,
        matchedDimensions: m.matchedDimensions,
        summary: m.memory.summary,
        skillUsed: m.memory.skillUsed,
        outcome: m.memory.outcome,
        expectedFiles: m.memory.changedFiles,
        clarifications: m.memory.clarifications,
      }));
      emitAndBroadcast(runId, "recall.matched", { queryEntities, matches: matchesPayload });
      recallMatchesForScan = matchesPayload;

      const { valid, stale } = validateRecalledFiles(repo, top.memory.changedFiles);
      if (stale.length > 0) {
        emitAndBroadcast(runId, "recall.stale", {
          historicalRunId: top.memory.runId,
          stale, valid,
          message: `${stale.length} file(s) from historical run no longer exist (likely refactored). Falling back to no-recall locate.`,
        });
        locateAttempt = 2;
      }
    }
  } catch (err) {
    console.warn(`[recall] 召回阶段失败（不阻塞）：${(err as Error).message}`);
  }

  // ── 2. Plan（前置到 user wait 之前 — 用 partial req 选 skill）──────────
  runPhase.set(runId, "plan");
  const ctx = repo.getContext();
  const planResult = await planAgent.run(analysis.req, ctx, llm);
  emitAndBroadcast(runId, "plan.done", {
    plan: planResult.plan,
    skillName: planResult.skill.name,
    score: planResult.score,
    by: planResult.by,
    candidates: planResult.candidates,
    routerReason: planResult.routerReason,
  });
  const skill = planResult.skill;

  // ── 2.5. aspectScan（WS-4 按需追问的核心）──────────
  runPhase.set(runId, "aspect-scan");
  const scanResult = await aspectScan(llm, {
    runId,
    rawText,
    candidateAspects: aspectUnion,
    recalledHistory: filterRecallForScan(recallMatchesForScan),
  });

  // 加固 2 末尾：过滤到属于选中 skill 的 aspect（跨 skill 同名 aspect 由 skill 内的语义决定）
  const skillAspectSet = new Set(skill.possibleAspects);
  const scanForSkill = scanResult.items.filter((it) => skillAspectSet.has(it.aspect));
  emitAndBroadcast(runId, "aspect.scanned", {
    skillName: skill.name,
    all: scanResult.items,
    forSkill: scanForSkill,
    breakdown: {
      explicit:     scanForSkill.filter((s) => s.status === "explicit").length,
      fromHistory:  scanForSkill.filter((s) => s.status === "from-history").length,
      needsAsking:  scanForSkill.filter((s) => s.status === "needs-asking").length,
    },
  });

  // ── 2.6. 步骤 B：对 needs-asking 的 aspect 拼问题（≥2 个时 LLM 打磨）──
  const questions = await buildQuestionsFromAspects(scanForSkill, skill, llm, runId);

  // ── 1.10. Emit clarify.questions（含 aspectScanResult 供前端审计）──
  emitAndBroadcast(runId, "clarify.questions", {
    questions,
    partial: analysis.req,
    aspectScan: scanForSkill,
  });

  // ── 3. Resolve（仅当有问题要等用户时）──
  runPhase.set(runId, "clarify");
  let req: ClarifiedRequest;
  if (questions.length > 0) {
    const answers = await waitForAnswers(runId);
    req = await clarifyAgent.resolve(rawText, analysis.req, answers, llm);
  } else {
    // 0 问题：直接把 partial + explicit/from-history 的 evidence 当 answers 写入
    const inferredAnswers: Record<string, string> = {};
    for (const item of scanForSkill) {
      if ((item.status === "explicit" || item.status === "from-history") && item.evidence) {
        inferredAnswers[item.aspect] = item.evidence;
      }
    }
    req = { ...analysis.req, answers: inferredAnswers };
  }
  emitAndBroadcast(runId, "clarify.done", { req });

  // ── 用 final req 重新算一遍 plan.files（partial 的 fieldName/businessRule 可能与 PM 实答略异）──
  const plan = await skill.plan(req, ctx);

  runPhase.set(runId, "locate");

  // ── 3. Locate ───────────────────────────────────────────
  let changes;
  try {
    changes = await locateAgent.run(skill, plan, ctx);
  } catch (err) {
    const e = err as { reason?: string; path?: string; message: string };
    emitAndBroadcast(runId, "locate.error", {
      reason: e.reason ?? "unknown",
      path: e.path,
      message: e.message,
    });
    throw err;
  }
  emitAndBroadcast(runId, "locate.done", {
    attempt: locateAttempt,
    files: changes.files,
    fileCount: changes.files.length,
  });
  runPhase.set(runId, "code");

  // ── 4. Code + Verify（最多重试 MAX_VERIFY_RETRIES 次）──
  let currentChanges: ChangeSet = changes;
  let patches = await runCodeAndVerify(runId, skill, currentChanges, llm, repo);

  // ── 5. Commit ───────────────────────────────────────────
  runPhase.set(runId, "commit");
  const branch = `feat/${req.fieldName}-${Date.now()}`;
  await repo.checkoutBranch(branch);
  const commitMsg = `feat: add ${req.fieldName} field\n\n${req.businessRule}`;
  await repo.stageAndCommit(commitMsg);

  const diff = await repo.getDiff("HEAD~1");
  const prDescription = await buildPRDescription(runId, req, diff);

  emitAndBroadcast(runId, "commit.done", {
    branch,
    commitMessage: commitMsg,
    prDescription,
    patchCount: patches.length,
  });

  emitAndBroadcast(runId, "run.completed", { runId });
  runPhase.delete(runId);

  // 异步沉淀 RequestMemory（不阻塞响应；失败仅 warn，不影响 run 状态）
  persistMemory(runId, llm)
    .then((mem) => {
      if (mem) console.log(`[memory] persisted runId=${runId} skill=${mem.skillUsed} entities=${JSON.stringify(mem.entities)}`);
    })
    .catch((err) => console.warn(`[memory] persist failed for runId=${runId}: ${(err as Error).message}`));
}

async function runCodeAndVerify(
  runId: string,
  skill: Skill,
  changes: ChangeSet,
  llm: ReturnType<typeof createDoubaoClient>,
  repo: ReturnType<typeof createConduitRepo>
): Promise<import("./types.js").FilePatch[]> {
  const repo_ = repo;
  let attempt = 1;

  while (true) {
    runPhase.set(runId, `code:attempt${attempt}`);
    // attempt 透传给 base.ts/runLLMOnFile，会进每条 llm.call 事件的 meta
    changes = { ...changes, meta: { ...changes.meta, attempt } };
    const patches = await codeAgent.run(skill, changes, llm, repo_);
    emitAndBroadcast(runId, "code.done", {
      attempt,
      files: patches.map((p) => p.path),
    });

    runPhase.set(runId, `verify:attempt${attempt}`);
    emitAndBroadcast(runId, "verify.running", { attempt });
    const result = await verifyAgent.run(repo_, patches.map((p) => p.path));

    if (result.success) {
      emitAndBroadcast(runId, "verify.done", { attempt, ...result });
      return patches;
    }

    emitAndBroadcast(runId, "verify.failed", {
      attempt,
      exitCode: result.exitCode,
      stderr: result.stderr.slice(0, 1000),
      stdout: result.stdout.slice(0, 1000),
    });

    if (attempt > MAX_VERIFY_RETRIES) {
      throw new Error(`verify 失败，已重试 ${MAX_VERIFY_RETRIES} 次。最后错误：\n${result.stderr.slice(0, 500)}`);
    }

    // 重试：把错误日志加入 meta，attempt + 1
    attempt++;
    changes = {
      ...changes,
      meta: {
        ...changes.meta,
        verifyError: result.stderr.slice(0, 2000),
        verifyStdout: result.stdout.slice(0, 2000),
      },
    };
    console.log(`[orchestrator] verify 失败，第 ${attempt} 次尝试...`);
  }
}

// ────────────────────────────────────────────────────────────
// 重放（从某事件之后丢弃，用新输入重跑）
// ────────────────────────────────────────────────────────────

export async function replayFrom(
  runId: string,
  fromEventIndex: number,
  newRawText: string
): Promise<void> {
  truncateAfter(runId, fromEventIndex);
  emitAndBroadcast(runId, "run.intervened", { fromEventIndex, newRawText });

  const llm = createDoubaoClient({ runId });
  const repo = createConduitRepo();

  runPipeline(runId, newRawText, llm, repo).catch((err) => {
    emitAndBroadcast(runId, "run.error", { message: String(err?.message ?? err) });
  });
}

// ────────────────────────────────────────────────────────────
// 查询
// ────────────────────────────────────────────────────────────

export function getRunEvents(runId: string): RunEvent[] {
  return readEvents(runId);
}

/** PM 在 RecallCard 点"忽略此条历史"时调用：只记事件、不改主流程（前端据此丢弃预填） */
export function dismissRecall(runId: string, historicalRunId: string): boolean {
  const events = readEvents(runId);
  if (!events.length) return false;
  emitAndBroadcast(runId, "recall.dismissed", { historicalRunId });
  return true;
}

export function getRunPhase(runId: string): string | undefined {
  return runPhase.get(runId);
}

File: packages/orchestrator/orchestrator/orchestrator.py
from __future__ import annotations

import asyncio
import time
from collections import defaultdict
from pathlib import Path
from typing import Any
from uuid import uuid4

from orchestrator.agents import code, code_spec, locate, plan_loop, verify
from orchestrator.agents.generic_plan import generic_skill
from orchestrator.events import store as event_store
from orchestrator.llm.doubao import create_doubao_client
from orchestrator.memory.extract import extract_entities
from orchestrator.memory.recall import recall
from orchestrator.memory.store import persist_memory, read_all_memories
from orchestrator.repo.conduit import ConduitRepo, create_conduit_repo
from orchestrator.skills.library_loader import load_skill_docs
from orchestrator.types import ChangeSet, RunEvent, dataclass_to_json

MAX_VERIFY_RETRIES = 2


class EventBus:
    def __init__(self) -> None:
        self._subscribers: dict[str, list[asyncio.Queue[RunEvent]]] = defaultdict(list)

    def emit(self, event: RunEvent) -> None:
        for q in list(self._subscribers.get(event.runId, [])):
            q.put_nowait(event)

    async def subscribe(self, run_id: str):
        q: asyncio.Queue[RunEvent] = asyncio.Queue()
        self._subscribers[run_id].append(q)
        try:
            yield q
        finally:
            self._subscribers[run_id].remove(q)


bus = EventBus()
pending_answers: dict[str, asyncio.Future[dict[str, str]]] = {}
run_phase: dict[str, str] = {}
run_repos: dict[str, ConduitRepo] = {}
run_tasks: dict[str, asyncio.Task] = {}


def emit_and_broadcast(run_id: str, event_type: str, payload: dict[str, Any]) -> RunEvent:
    event = event_store.emit(run_id, event_type, payload)
    bus.emit(event)
    return event


def _branch_slug(title: str) -> str:
    slug = "".join(c if (c.isascii() and c.isalnum()) else "-" for c in title.strip().lower())
    slug = "-".join(part for part in slug.split("-") if part)
    return slug[:40].strip("-") or "change"


def _plan_payload_files(files) -> list[dict[str, Any]]:
    return [dataclass_to_json(f) for f in files]


async def start_run(raw_text: str, repo_url: str | None = None) -> str:
    run_id = str(uuid4())
    llm = create_doubao_client({"runId": run_id})
    repo = ConduitRepo.from_url(repo_url) if repo_url else create_conduit_repo()
    run_repos[run_id] = repo
    task = asyncio.create_task(_run_pipeline(run_id, raw_text, llm, repo))
    run_tasks[run_id] = task
    return run_id


def provide_clarification_answers(run_id: str, answers: dict[str, str]) -> bool:
    fut = pending_answers.pop(run_id, None)
    if not fut or fut.done():
        return False
    fut.set_result(answers)
    return True


def stop_run(run_id: str) -> bool:
    task = run_tasks.get(run_id)
    if not task or task.done():
        return False

    fut = pending_answers.pop(run_id, None)
    if fut and not fut.done():
        fut.cancel()

    task.cancel()
    return True


def is_run_active(run_id: str) -> bool:
    task = run_tasks.get(run_id)
    return bool(task and not task.done())


async def _wait_for_answers(run_id: str) -> dict[str, str]:
    fut: asyncio.Future[dict[str, str]] = asyncio.get_running_loop().create_future()
    pending_answers[run_id] = fut
    return await fut


async def _run_pipeline(run_id: str, raw_text: str, llm, repo) -> None:
    try:
        emit_and_broadcast(
            run_id,
            "run.started",
            {
                "rawText": raw_text,
                "recallDisabled": False,
                "repoPath": str(repo.repo_path),
                "repoNwo": repo.upstream_nwo,
            },
        )
        # ---- recall：检索相似历史需求，作为规划知识源 ----
        run_phase[run_id] = "recall"
        recall_matches: list[dict[str, Any]] = []
        try:
            query_entities = await extract_entities(llm, run_id=run_id, summary=raw_text, skill_used="")
            repo_nwo = run_repos.get(run_id, None)
            repo_nwo_str = repo_nwo.upstream_nwo if repo_nwo else None
            matches = recall(query_entities, read_all_memories(), top_k=3, min_score=0.3, repo_nwo=repo_nwo_str)
            if matches:
                recall_matches = [
                    {
                        "runId": m.memory.runId,
                        "score": m.score,
                        "matchedDimensions": m.matchedDimensions,
                        "summary": m.memory.summary,
                        "skillUsed": m.memory.skillUsed,
                        "outcome": m.memory.outcome,
                        "expectedFiles": m.memory.changedFiles,
                        "clarifications": [c.__dict__ for c in m.memory.clarifications],
                    }
                    for m in matches
                ]
                emit_and_broadcast(run_id, "recall.matched", {"queryEntities": query_entities.__dict__, "matches": recall_matches})
        except Exception:
            pass

        # ---- code-spec：建/更新索引，算出相关文件摘要 ----
        run_phase[run_id] = "code-spec"
        ctx = repo.get_context()
        code_spec_context = None
        try:
            spec_status = await code_spec.ensure_current(repo, llm, run_id)
            code_spec_context = code_spec.format_relevant_context(raw_text, repo)
            ctx.codeSpecContext = code_spec_context
            emit_and_broadcast(run_id, "spec.ready", spec_status)
        except Exception as e:
            emit_and_broadcast(run_id, "spec.failed", {"message": str(e)})

        # ---- plan-loop：规划与澄清合一。plan agent 自评信息是否足够，不够才提问，最多 N 轮，到顶强制出规划 ----
        run_phase[run_id] = "plan"
        skills_context = load_skill_docs()
        recall_context = plan_loop.format_recall_context(recall_matches)
        qa_history: list[tuple[str, str]] = []
        plan_out = None
        for round_idx in range(plan_loop.MAX_QUESTION_ROUNDS + 1):
            force_ready = round_idx == plan_loop.MAX_QUESTION_ROUNDS
            plan_out = await plan_loop.plan_step(
                raw_text,
                ctx,
                llm,
                skills_context=skills_context,
                code_spec_context=code_spec_context,
                recall_context=recall_context,
                qa_history=qa_history,
                force_ready=force_ready,
            )
            if plan_out.status == "ready":
                break
            run_phase[run_id] = "clarify"
            emit_and_broadcast(
                run_id,
                "clarify.questions",
                {
                    "questions": plan_out.questions,
                    "partial": {"summary": raw_text},
                    "aspectScan": [],
                    "mode": "generic",
                    "round": round_idx + 1,
                },
            )
            answers = await _wait_for_answers(run_id)
            for q in plan_out.questions:
                qa_history.append((q, answers.get(q, "")))
            run_phase[run_id] = "plan"

        final_plan = plan_out.plan
        plan_title = plan_out.title or (raw_text[:48].strip() or "change")

        emit_and_broadcast(
            run_id,
            "clarify.done",
            {"req": {"summary": raw_text, "answers": {q: a for q, a in qa_history}}},
        )
        emit_and_broadcast(
            run_id,
            "plan.generic",
            {
                "mode": "generic",
                "plan": dataclass_to_json(final_plan),
                "files": _plan_payload_files(final_plan.files),
                "score": 0,
                "by": "plan-loop",
                "candidates": [],
                "routerReason": None,
                "reason": "plan-loop（规划与澄清合一）",
            },
        )

        skill = generic_skill
        run_phase[run_id] = "locate"
        changes = await locate.run(skill, final_plan, ctx)
        emit_and_broadcast(run_id, "locate.done", {"attempt": 1, "files": [f.__dict__ for f in changes.files], "fileCount": len(changes.files)})
        patches = await _run_code_and_verify(run_id, skill, changes, llm, repo)

        run_phase[run_id] = "code-spec:update"
        try:
            spec_update = await code_spec.update_files(repo, llm, [p.path for p in patches], run_id)
            emit_and_broadcast(run_id, "spec.updated", spec_update)
        except Exception as e:
            emit_and_broadcast(run_id, "spec.update_failed", {"message": str(e)})

        run_phase[run_id] = "commit"
        branch = f"feat/{_branch_slug(plan_title)}-{int(time.time() * 1000)}"
        repo.checkout_branch(branch)
        commit_msg = f"feat: {plan_title}\n\n{raw_text}"
        repo.stage_and_commit(commit_msg, [p.path for p in patches])

        run_phase[run_id] = "pr"
        pr_title = f"feat: {plan_title}"
        pr_body = raw_text or commit_msg
        repo.push_branch(branch)
        pr_url = repo.create_pr(branch, pr_title, pr_body)

        emit_and_broadcast(
            run_id,
            "commit.done",
            {"branch": branch, "commitMessage": commit_msg, "prDescription": pr_body, "patchCount": len(patches), "prUrl": pr_url},
        )
        emit_and_broadcast(run_id, "run.completed", {"runId": run_id})
        run_phase.pop(run_id, None)
        try:
            _repo = run_repos.get(run_id)
            await persist_memory(run_id, llm, repo_nwo=_repo.upstream_nwo if _repo else None)
        except Exception:
            pass
        run_repos.pop(run_id, None)
    except asyncio.CancelledError:
        emit_and_broadcast(
            run_id,
            "run.cancelled",
            {
                "phase": run_phase.get(run_id, "unknown"),
                "message": "Run stopped by user. Generated files may remain in the target worktree.",
            },
        )
        raise
    except Exception as e:
        emit_and_broadcast(run_id, "run.error", {"message": str(e)})
    finally:
        run_phase.pop(run_id, None)
        pending_answers.pop(run_id, None)
        run_tasks.pop(run_id, None)
        run_repos.pop(run_id, None)


async def _run_code_and_verify(run_id: str, skill, changes: ChangeSet, llm, repo) -> list:
    attempt = 1
    while True:
        run_phase[run_id] = f"code:attempt{attempt}"
        changes.meta = {**changes.meta, "attempt": attempt}
        patches = await code.run(skill, changes, llm, repo)
        emit_and_broadcast(run_id, "code.done", {"attempt": attempt, "files": [p.path for p in patches]})
        run_phase[run_id] = f"verify:attempt{attempt}"
        emit_and_broadcast(run_id, "verify.running", {"attempt": attempt})
        result = await verify.run(repo, [p.path for p in patches])
        if result.success:
            emit_and_broadcast(run_id, "verify.done", {"attempt": attempt, **result.__dict__})
            return patches
        emit_and_broadcast(run_id, "verify.failed", {"attempt": attempt, **result.__dict__})
        if attempt > MAX_VERIFY_RETRIES:
            raise RuntimeError(f"verify 失败，已重试 {MAX_VERIFY_RETRIES} 次")
        attempt += 1
        changes.meta = {**changes.meta, "verifyError": result.stderr[:2000], "verifyStdout": result.stdout[:2000]}
        # 让下一次重试看到自己上一次生成的（未通过的）内容，而不是 locate 时抓的原始内容
        changes.context = {**changes.context, **{p.path: p.newContent for p in patches}}


async def replay_from(run_id: str, from_event_index: int, new_raw_text: str) -> None:
    event_store.truncate_after(run_id, from_event_index)
    emit_and_broadcast(run_id, "run.intervened", {"fromEventIndex": from_event_index, "newRawText": new_raw_text})
    llm = create_doubao_client({"runId": run_id})
    repo = run_repos.get(run_id)
    if repo is None:
        started = next((e for e in get_run_events(run_id) if e.type == "run.started"), None)
        repo_path = ((started.payload or {}).get("repoPath") if started else None) if started else None
        repo_nwo = ((started.payload or {}).get("repoNwo") if started else None) if started else None
        repo = ConduitRepo(Path(repo_path), upstream_nwo=repo_nwo) if repo_path else create_conduit_repo()
    run_repos[run_id] = repo
    task = asyncio.create_task(_run_pipeline(run_id, new_raw_text, llm, repo))
    run_tasks[run_id] = task


async def resume_run(run_id: str) -> bool:
    active = run_tasks.get(run_id)
    if active and not active.done():
        return False

    events = get_run_events(run_id)
    if not events:
        return False

    raw_text = ""
    for event in reversed(events):
        if event.type == "run.started":
            raw_text = str(event.payload.get("rawText") or "")
            break
    if not raw_text:
        return False

    safe_types = {"run.started", "clarify.done", "plan.done", "plan.generic", "locate.done", "code.done", "verify.done"}
    keep_count = 0
    for index, event in enumerate(events):
        if event.type in safe_types:
            keep_count = index + 1

    await replay_from(run_id, keep_count, raw_text)
    return True


def get_run_events(run_id: str) -> list[RunEvent]:
    return event_store.read_events(run_id)


def dismiss_recall(run_id: str, historical_run_id: str) -> bool:
    if not get_run_events(run_id):
        return False
    emit_and_broadcast(run_id, "recall.dismissed", {"historicalRunId": historical_run_id})
    return True


def get_run_phase(run_id: str) -> str | None:
    return run_phase.get(run_id)
